### ISSUE TITLE
Canvas Context : setLineDash with empty array provoke exception

### ISSUE THREAD

**Issue by chourie2** [wkhtmltopdf/wkhtmltopdf#3938]
Hello,

This issue is related to issue #2353 that was fixed in 0.12.4. My OS is Windows 10 version 1709. 

I encounter an error (with application crash) when calling the function setLineDash with an empty array parameter to draw a solid line without dash. Such call leads probably to an null pointer exception on the call of context.setLineDash (line 184 of JSCanvasRenderingContext2DCustom.cpp). 

Based on my interpretation of the documentation ([here](https://developer.mozilla.org/en-US/docs/Web/API/CanvasRenderingContext2D/setLineDash) or [there](https://www.w3.org/TR/2dcontext/#dom-context-2d-setlinedash)) I would say that it's suppose to work, and it actually does in others browsers.

You can reproduce this issue with this simple use case by using wkhtmltopdf whitout any option : [SetLineDash.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2053053/SetLineDash.zip)

Thanks in advance for your time, I'll stay tuned if you need anything else ! 😄

**Comment by ashkulz** [comment]
So, thanks for the detailed analysis. If nothing is done when an empty array is passed would it work? Because this patch would be obsolete once the WebKit version is upgraded (as it is a custom backport).

**Comment by chourie2** [comment]
I think your suggestion would only prevent the crash, but not draw the line. That's still an improvement, but it might be possible to draw the line without recoding everything !

**Comment by ashkulz** [comment]
So there's a PR with what I suggested already wkhtmltopdf/qt#35 -- might want to check it.

**Comment by chourie2** [comment]
I wasn't able to build on Windows, I'm not sure that I was doing it right since this is kind of a work in progress! Just in case, I encounter an error during the nmake command of the QT build, something about hash_map being deprecated. I won't elaborate too much since I read that you where working on the packaging, but I can create a new issue if needed.

However, I would be happy to test it if @quique or someone else would be kind enough to provide me a build result (.exe for windows MSVC 2015) from this PR.

But I was certainly wrong on my last comment, this seems to be the correct fix for both quique issue and mine (which we encounter when using 2 related js library).

**Comment by chourie2** [comment]
Hello,

I've seen your merge commit from quique's PR, and I just use the 0.12.6 snapshot to finally test it! The good news is that there's no more exception, so that's great!

The bad news is that it doesn't draw a solid line 😢 

Here's the result : [SetLineDash.pdf](https://github.com/wkhtmltopdf/wkhtmltopdf/files/4109608/SetLineDash.pdf)

The second line was supposed to be solid, without any dash, like we can see in the html original file (see [SetLineDash.zip](https://github.com/wkhtmltopdf/wkhtmltopdf/files/2053053/SetLineDash.zip))

Have you any idea how to fix that, or can you make some test, or help me build wkhtmltopdf so I can make them by myself ?

**Comment by ashkulz** [comment]
I don't really know, sorry. I do recall someone mentioning a workaround, you might want to search through the issues for it.

**Comment by chourie2** [comment]
Yes, I found it : https://github.com/wkhtmltopdf/wkhtmltopdf/issues/3242#issuecomment-518099192
I think the same implementation in wkhtmltopdf would be more appropriate regarding the HTML standard. 

In the meantime I can indeed use this workaround! Thank you

**Comment by joshuapinter** [comment]
For what it's worth we were running into this when we upgraded from 0.12.3 to 0.12.5 and upgrading to 0.12.6, the issue went away. We believe it had to do with our use of `jCanvas` and its use of `setLineDash`.
